### ISSUE TITLE
Enable providers to transfer 64 bit integers

### ISSUE THREAD

**Issue by iwahbe** [pulumi/pulumi#14242]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Providers ([pulumi-libvert](https://github.com/pulumi/pulumi-libvirt/issues/6)) expect to be able to transfer 64 bit integers across the wire. Bridged providers expect at least system width integers (64 bit on a 64 bit processor). Not all SDKs support this. For example, C#'s `int` is 32 bits, regardless of the processor.

It seems like the Pulumi type system will need to intrude a `LongType`, since `IntType` is labeled as a 32 bit value:

https://github.com/pulumi/pulumi/blob/4ebf61f896fe226bf455da3fc87c91abdc33dc28/pkg/codegen/schema/schema.go#L92-L93

To prevent breaking changes, it might be necessary for fields to opt in to 64 bit values in the pulumi schema.

**Comment by justinvp** [comment]
In addition to the schema + codegen changes, fully supporting this requires updates to the state and wire protocols, which currently encodes numbers as floats (like JSON). Integer support today is done via casting to/from float in SDKs. https://github.com/pulumi/pulumi/pull/13868 explores adding integer support.

**Comment by t0yv0** [comment]
64-bit integers are not sufficient as Terraform providers permit big.Float in theory.

**Comment by iwahbe** [comment]
> Also it appears we started deploying workarounds in the bridge where things are encoded as `string` when they overflow, but it has not universally fixed the problems yet.

I think this works only up to int64, I don't know if it would work to support `*big.Float`. The major problem for the bridge work-arounds is that they need to be reactive. We only can fix a problem once a user reports it, and the fix is breaking.

**Comment by Frassle** [comment]
When I looked into this it was easier for us to just support bigint rather than int64 anyway.

**Comment by t0yv0** [comment]
Yeah the rollout is not easy. We could thread the breaking changes through as major versions. But also identifying locations where this is needed is tricky. I doubt we want all ints/floats to become big.Int in the SDK as in many places the precision does not matter. For now we can just accumulate issues with impact/breaking and cross-linked from here.
